fmt/src/formatter.rs

```diff
@@ -0,0 +1,561 @@
+            let is_pragma =
+                |u: &SourceUnitPart| matches!(u, SourceUnitPart::PragmaDirective(_, _, _, _));
+            let is_import = |u: &SourceUnitPart| matches!(u, SourceUnitPart::ImportDirective(_, _));
+            let is_declaration = |u: &SourceUnitPart| !(is_pragma(u) || is_import(u));
+
+            unit.visit(self)?;
+            writeln!(self)?;
+
+            if let Some(next_unit) = source_unit_parts_iter.peek() {
+                if (is_declaration(unit) || is_declaration(next_unit)) ||
+                    (is_pragma(unit) || is_pragma(next_unit)) ||
+                    (is_import(unit) &&
+                        is_import(next_unit) &&
+                        // If source has zero blank lines between imports, leave it as is. If one
+                        //  or more, separate imports with one blank line.
+                        self.blank_lines(unit, next_unit) > 1)
+                {
+                    writeln!(self)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn visit_doc_comment(&mut self, doc_comment: &mut DocComment) -> VResult {
+        match doc_comment {
+            DocComment::Line { comment } => {
+                write!(self, "/// @{}", comment.tag)?;
+                if !comment.value.is_empty() {
+                    let mut lines = comment.value.split('\n');
+                    write!(self, " {}", lines.next().unwrap())?;
+
+                    for line in lines {
+                        writeln!(self)?; // Write newline separately to trigger an indentation
+                        write!(self, "/// {}", line)?;
+                    }
+                }
+            }
+            DocComment::Block { comments } => {
+                writeln!(self, "/**")?;
+                for comment in comments {
+                    write!(self, "@{} ", comment.tag)?;
+                    for line in comment.value.split('\n') {
+                        writeln!(self, "{}", line)?;
+                    }
+                }
+                write!(self, "*/")?;
+            }
+        };
+
+        Ok(())
+    }
+
+    fn visit_doc_comments(&mut self, doc_comments: &mut [DocComment]) -> VResult {
+        for (i, doc_comment) in doc_comments.iter_mut().enumerate() {
+            if i > 0 {
+                writeln!(self)?;
+            }
+            doc_comment.visit(self)?;
+        }
+
+        Ok(())
+    }
+
+    fn visit_contract(&mut self, contract: &mut ContractDefinition) -> VResult {
+        if !contract.doc.is_empty() {
+            contract.doc.visit(self)?;
+            writeln!(self)?;
+        }
+
+        write!(self, "{} {} ", contract.ty, contract.name.name)?;
+
+        if !contract.base.is_empty() {
+            write!(self, "is")?;
+
+            let bases = contract
+                .base
+                .iter_mut()
+                .map(|base| {
+                    // TODO
+                    self.visit_to_string(&mut base.loc)
```

Comment:
what's the todo here?

Reply:
oops, forgot to add explainer. It was intended as 
```rust
// TODO: implement `Visitable` for `Base` and visit it here instead of its loc
```